src/FreeAge/common/player.hpp

```diff
@@ -135,3 +136,36 @@ struct PlayerStats {
   /// If a building has ever been constructed per building type.
   bool buildingExisted[static_cast<int>(BuildingType::NumBuildings)];
 };
+
+struct Player {
```

Comment:
@puzzlepaint Take look at this. Basically have this base Player class for both the server and the client for keeping track all the type stats. Server will have all the real values and client will have the current players real values and the latest know for each of the other players. Then access to the stats would be 
```c++
players.at(playerIndex).GetUnitStats(UnitType::Scout).fireRate
```
Or this could be wrapped in the Client/SeverUnit/Building
```c++
unit->GetStats().fireRate
```
Also, you were right (again), there no need for a UnitTypeInfo. The AddModification could be added inside the Player and apply the effects directly to the data in the vectors (which is client/server independent).

Reply:
Sounds / looks good to me. A shortcut such as `unit->GetStats().fireRate` (perhaps not in the Unit class though as this might not have access to `players`) is probably highly desirable, since I guess that in many cases, the stats for some specific object are needed, and it would be quite cumbersome to always do `players.at(unit->GetPlayerIndex()).GetUnitStats(unit->GetType()).fireRate`.

Reply:
Unit has a playerIndex, shouldn't be replaced with a `Player*`? It should be safe to use a pointer.

Reply:
Indeed, this seems like it should solve this nicely.